File: apps/api/src/f1hub/api/v1/live.py
"""
Live Streaming API Endpoints

FastAPI endpoints for live F1 data streaming.
WebSocket functionality will be added later.
"""

from typing import Optional
from uuid import UUID
import logging

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session as DBSession

from ...core.dependencies import get_db
from ...db.models import LiveEvent, LiveSession, LiveTiming, Session

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/live", tags=["live"])


@router.post("/sessions/start", status_code=status.HTTP_201_CREATED)
def start_live_session(
    session_id: UUID,
    openf1_session_key: str,
    db: DBSession = Depends(get_db),
):
    """
    Start a live streaming session.

    Args:
        session_id: F1Hub session UUID
        openf1_session_key: OpenF1 API session key

    Returns:
        LiveSession: Created live session
    """
    # Verify session exists
    session = db.query(Session).filter(Session.id == session_id).first()
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")

    # Check if live session already exists
    existing = db.query(LiveSession).filter(
        LiveSession.session_id == session_id,
        LiveSession.is_active == True
    ).first()

    if existing:
        return {
            "id": str(existing.id),
            "session_id": str(existing.session_id),
            "is_active": existing.is_active,
            "openf1_session_key": existing.openf1_session_key,
            "current_lap": existing.current_lap,
            "session_status": existing.session_status,
            "track_status": existing.track_status,
        }

    # Create new live session
    live_session = LiveSession(
        session_id=session_id,
        openf1_session_key=openf1_session_key,
        is_active=True,
        session_status="Started",
        track_status="Unknown",
    )

    db.add(live_session)
    db.commit()
    db.refresh(live_session)

    logger.info(f"Started live session {live_session.id} for session {session_id}")

    return {
        "id": str(live_session.id),
        "session_id": str(live_session.session_id),
        "is_active": live_session.is_active,
        "openf1_session_key": live_session.openf1_session_key,
        "current_lap": live_session.current_lap,
        "session_status": live_session.session_status,
        "track_status": live_session.track_status,
    }


@router.post("/sessions/{live_session_id}/stop")
def stop_live_session(
    live_session_id: UUID,
    db: DBSession = Depends(get_db),
):
    """
    Stop a live streaming session.

    Args:
        live_session_id: LiveSession UUID

    Returns:
        Success message
    """
    live_session = db.query(LiveSession).filter(LiveSession.id == live_session_id).first()
    if not live_session:
        raise HTTPException(status_code=404, detail="Live session not found")

    live_session.is_active = False
    live_session.ended_at = None  # Will be set by database default
    db.commit()

    logger.info(f"Stopped live session {live_session_id}")

    return {"message": "Live session stopped", "live_session_id": str(live_session_id)}


@router.get("/sessions/active")
def get_active_sessions(db: DBSession = Depends(get_db)):
    """
    Get all active live sessions.

    Returns:
        List of active LiveSession records
    """
    sessions = db.query(LiveSession).filter(LiveSession.is_active == True).all()

    return [
        {
            "id": str(session.id),
            "session_id": str(session.session_id),
            "is_active": session.is_active,
            "openf1_session_key": session.openf1_session_key,
            "current_lap": session.current_lap,
            "session_status": session.session_status,
            "track_status": session.track_status,
            "started_at": session.created_at.isoformat() if session.created_at else None,
        }
        for session in sessions
    ]


@router.get("/sessions/{live_session_id}/timing")
def get_live_timing(
    live_session_id: UUID,
    limit: int = 100,
    db: DBSession = Depends(get_db),
):
    """
    Get recent timing data for a live session.

    Args:
        live_session_id: LiveSession UUID
        limit: Maximum number of records to return

    Returns:
        List of LiveTiming records
    """
    timing_data = (
        db.query(LiveTiming)
        .filter(LiveTiming.live_session_id == live_session_id)
        .order_by(LiveTiming.timestamp.desc())
        .limit(limit)
        .all()
    )

    return [
        {
            "driver_number": t.driver_number,
            "lap_number": t.lap_number,
            "position": t.position,
            "gap_to_leader": t.gap_to_leader,
            "interval": t.interval,
            "last_lap_time": t.last_lap_time,
            "sector1_time": t.sector1_time,
            "sector2_time": t.sector2_time,
            "sector3_time": t.sector3_time,
            "timestamp": t.timestamp.isoformat() if t.timestamp else None,
        }
        for t in timing_data
    ]


@router.get("/sessions/{live_session_id}/events")
def get_live_events(
    live_session_id: UUID,
    event_type: Optional[str] = None,
    limit: int = 100,
    db: DBSession = Depends(get_db),
):
    """
    Get events for a live session.

    Args:
        live_session_id: LiveSession UUID
        event_type: Filter by event type (pit_stop, race_control, etc.)
        limit: Maximum number of records to return

    Returns:
        List of LiveEvent records
    """
    query = db.query(LiveEvent).filter(LiveEvent.live_session_id == live_session_id)

    if event_type:
        query = query.filter(LiveEvent.event_type == event_type)

    events = query.order_by(LiveEvent.timestamp.desc()).limit(limit).all()

    return [
        {
            "id": str(e.id),
            "event_type": e.event_type,
            "driver_number": e.driver_number,
            "lap_number": e.lap_number,
            "description": e.description,
            "severity": e.severity,
            "data": e.data,
            "timestamp": e.timestamp.isoformat() if e.timestamp else None,
        }
        for e in events
    ]

File: apps/web/src/lib/api/live.ts
/**
 * Live Streaming API Client
 *
 * Client functions for live F1 data streaming endpoints.
 */

import { apiClient } from './client';

export interface LiveSession {
  id: string;
  session_id: string;
  is_active: boolean;
  openf1_session_key: string;
  current_lap: number | null;
  session_status: string;
  track_status: string;
  started_at?: string;
}

export interface LiveTiming {
  driver_number: number;
  lap_number: number;
  position: number;
  gap_to_leader: number | null;
  interval: number | null;
  last_lap_time: number | null;
  sector1_time: number | null;
  sector2_time: number | null;
  sector3_time: number | null;
  timestamp: string;
}

export interface LiveEvent {
  id: string;
  event_type: string;
  driver_number: number | null;
  lap_number: number | null;
  message: string;
  flag: string | null;
  data: Record<string, any>;
  timestamp: string;
}

export interface LiveUpdate {
  type: 'live_update';
  data: {
    timing: any[];
    pit_stops: any[];
    race_control: any[];
    session_status: any;
    weather: any;
  };
  timestamp: string;
  session_id: string;
}

/**
 * Start a live streaming session
 */
export async function startLiveSession(
  sessionId: string,
  openf1SessionKey: string
): Promise<LiveSession> {
  const response = await apiClient.post<LiveSession>(
    `/live/sessions/start`,
    null,
    {
      params: {
        session_id: sessionId,
        openf1_session_key: openf1SessionKey,
      },
    }
  );
  return response.data;
}

/**
 * Stop a live streaming session
 */
export async function stopLiveSession(
  liveSessionId: string
): Promise<{ message: string; live_session_id: string }> {
  const response = await apiClient.post(
    `/live/sessions/${liveSessionId}/stop`
  );
  return response.data;
}

/**
 * Get all active live sessions
 */
export async function getActiveLiveSessions(): Promise<LiveSession[]> {
  const response = await apiClient.get<LiveSession[]>('/live/sessions/active');
  return response.data;
}

/**
 * Get live timing data for a session
 */
export async function getLiveTiming(
  liveSessionId: string,
  limit: number = 100
): Promise<LiveTiming[]> {
  const response = await apiClient.get<LiveTiming[]>(
    `/live/sessions/${liveSessionId}/timing`,
    {
      params: { limit },
    }
  );
  return response.data;
}

/**
 * Get live events for a session
 */
export async function getLiveEvents(
  liveSessionId: string,
  eventType?: string,
  limit: number = 100
): Promise<LiveEvent[]> {
  const response = await apiClient.get<LiveEvent[]>(
    `/live/sessions/${liveSessionId}/events`,
    {
      params: { event_type: eventType, limit },
    }
  );
  return response.data;
}

/**
 * Get WebSocket connection count
 */
export async function getConnectionCount(
  sessionId?: string
): Promise<{ count: number; session_id: string | null }> {
  const response = await apiClient.get<{ count: number; session_id: string | null }>(
    '/live/connections/count',
    {
      params: sessionId ? { session_id: sessionId } : {},
    }
  );
  return response.data;
}
